js/core/option_manager.js

```diff
@@ -0,0 +1,215 @@
+import coreDataUtils from "./utils/data";
+import typeUtils from "./utils/type";
+import compareUtils from "./utils/compare";
+import CallBacks from "./utils/callbacks";
+import { extend } from "./utils/extend";
+
+export class OptionManager {
+    constructor(options, optionsByReference, deprecatedOptions) {
+        this._options = options;
+        this._optionsByReference = optionsByReference;
+        this._deprecatedOptions = deprecatedOptions;
+        this._changingCallbacks = CallBacks({ syncStrategy: true });
+        this._changedCallbacks = CallBacks({ syncStrategy: true });
+        this._deprecatedCallbacks = CallBacks({ syncStrategy: true });
+        this._cachedDeprecateNames = [];
+        this.cachedGetters = {};
+        this.cachedSetters = {};
+    }
+
+    _logWarningIfDeprecated(option) {
+        const info = this._deprecatedOptions[option];
+        if(info) {
+            this._deprecatedCallbacks.fire(option, info);
+        }
+    }
+
+    _clearField(options, name) {
+        delete options[name];
+
+        const previousFieldName = this._getParentName(name);
+        const fieldName = this._getFieldName(name);
+        const fieldObject = previousFieldName ? this._getValue(options, previousFieldName, false) : options;
+
+        if(fieldObject) {
+            delete fieldObject[fieldName];
```

Comment:
No need to have a variable
```suggestion
            delete fieldObject[this._getFieldName(name)];
```